### PR TITLE
[IPU] update ipu_backend p0

### PR DESCRIPTION
**Pull request by gglin001**
### PR types

Others
### PR changes

Others
### Describe

update ipu_backend
check p1 at https://github.com/PaddlePaddle/Paddle/pull/38986

### CODE REVIEW

File: paddle/fluid/platform/device/ipu/CMakeLists.txt [hunk 1]
```diff
@@ -1,12 +1,31 @@
...
-  cc_library(ipu_compiler SRCS ipu_compiler.cc DEPS popart graph ipu_utils graph_helper)
-  cc_library(ipu_backend SRCS ipu_backend.cc DEPS popart ipu_compiler graph framework_proto enforce ipu_utils ipu_strategy ipu_device ipu_executor graph_helper)
+
+  set(POPART_PATTERNS_SRC
+    ""
+  )
+
+  set(POPART_OPERATORS_SRC
+    ""
+  )
```
Comment (qili93) [inline paddle/fluid/platform/device/ipu/CMakeLists.txt]: "POPART_PATTERNS_SRC" 和 "POPART_OPERATORS_SRC" 修改之后为空了，建议直接删掉，不留冗余代码

File: paddle/fluid/platform/device/ipu/ipu_device.cc [hunk 2]
```diff
@@ -0,0 +1,55 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qili93) [inline paddle/fluid/platform/device/ipu/ipu_device.cc]: 时间修改一下，新增文件时间改为2022

File: paddle/fluid/platform/device/ipu/ipu_executor.cc [hunk 3]
```diff
@@ -78,96 +71,116 @@ void Executor::Prepare(const std::string &proto,
...
-  if (ipu_strategy_ != nullptr && ipu_strategy_->is_training) {
+  step_++;
+  if (ipu_strategy_->is_training &&
+      step_ % ipu_strategy_->save_per_n_step == 0) {
     session_->weightsToHost();
     WeightsToPaddle();
-    if (ipu_strategy_->save_last_onnx) {
-      session_->modelToHost("test_last.onnx");
+    if (ipu_strategy_->save_onnx_checkpoint) {
+      session_->modelToHost("test_last" + std::to_string(step_) + ".onnx");
```
Comment (qili93) [inline paddle/fluid/platform/device/ipu/ipu_executor.cc]: 建议这里增加对Paddle paddle.Model.save保存paddle模型格式的checkpoint的支持

File: paddle/fluid/platform/device/ipu/ipu_utils.h [hunk 4]
```diff
@@ -17,14 +17,27 @@ limitations under the License. */
...
 namespace ipu {
 
+using platform::float16;
+using framework::Tensor;
+using framework::LoDTensor;
+using framework::Scope;
+using framework::OpDesc;
+using framework::ir::Graph;
+using framework::ir::Node;
+using framework::BlockDesc;
```
Comment (qili93) [inline paddle/fluid/platform/device/ipu/ipu_utils.h]: 从32-39行代码都建议写成 using float16 = platform::float16; 这样的方式

File: paddle/fluid/platform/device/ipu/ipu_device.cc (hunk 2 above)
Comment (gglin001) [reply paddle/fluid/platform/device/ipu/ipu_device.cc]: ok, 包括其他 PR 有关时间的修改都会在后面更新

File: paddle/fluid/platform/device/ipu/CMakeLists.txt (hunk 1 above)
Comment (gglin001) [reply paddle/fluid/platform/device/ipu/CMakeLists.txt]: solved

File: paddle/fluid/platform/device/ipu/ipu_executor.cc (hunk 3 above)
Comment (gglin001) [reply paddle/fluid/platform/device/ipu/ipu_executor.cc]: 我们已经支持了通过调用 `paddle.static.save` 保存  checkpoint

File: paddle/fluid/platform/device/ipu/ipu_utils.h (hunk 4 above)
Comment (gglin001) [reply paddle/fluid/platform/device/ipu/ipu_utils.h]: fixed

**Comment by qili93** [comment]
PR-CI-APPROVAL 提示需要修改

2022-01-12 10:49:52 1. The error message you wrote in PADDLE_ENFORCE{_**} or PADDLE_THROW does not meet our error message writing specification. Possible errors include 1. the error message is empty / 2. the error message is too short / 3. the error type is not specified. Please read the specification [ https://github.com/PaddlePaddle/Paddle/wiki/Paddle-Error-Message-Writing-Specification ], then refine the error message. If it is a mismatch, please request chenwhql (Recommend), luotao1 or lanxianghit review and approve.
2022-01-12 10:49:52 The PADDLE_ENFORCE{_**} or PADDLE_THROW entries that do not meet the specification are as follows:
2022-01-12 10:49:52 PADDLE_THROW(platform::errors::InvalidArgument("Uknown AdamMode: %s", str)); 
2022-01-12 10:49:52 PADDLE_THROW( + platform::errors::NotFound("%s is not registered", op_type)); 
2022-01-12 10:49:52 PADDLE_ENFORCE_NE(tensor, resources_->tensors.end(), + resources_->outputs.push_back(tensor->second);

File: paddle/fluid/platform/device/ipu/supported_ops_custom.h [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// clang-format off
+
+#pragma once
+
+OP_DECL(popart_nllloss_v2, aiGraphcoreOpset.nllloss, SIG_ARG(INT32,popart::ReductionType,reduction) OPT_ARG(INT32,ignoreIndex) ARG(BOOL,inputIsLogProbability) ) // NOLINT
```
Comment (qili93) [inline paddle/fluid/platform/device/ipu/supported_ops_custom.h:19]: // clang-format off 之后，在文件结尾都需要加上 // clang-format on

**Review by qili93 — APPROVED** [review]
LGTM

**Pull request merged**